simulation/halsim_ws_core/src/main/native/cpp/WSHalProviders.cpp

```diff
@@ -0,0 +1,55 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "WSHalProviders.h"
+
+namespace wpilibws {
+
+void HALSimWSHalProvider::OnNetworkConnected(
+    std::shared_ptr<HALSimBaseWebSocketConnection> ws) {
+  {
+    std::lock_guard lock(mutex);
+
+    // store a weak reference to the websocket object
+    m_ws = ws;
+  }
+
+  // trigger a send of the current state
+  OnSimCallback("");
+}
+
+void HALSimWSHalProvider::OnStaticSimCallback(const char* name, void* param,
+                                              const struct HAL_Value* value) {
```

Comment:
This HAL_Value is what I was referring to.

Could probably just make a generic function that takes a param that has the base json value already populated? Then you make a copy of the json blob, fill in the value from the HAL_Value, and send the data off to the websocket. Then we could cut out a ton of code.

Reply:
Ugh. If I had done it that way originally this would have been so much simpler.

Reply:
We might still end up repeating a lot of code in order to generate the full-state message whenever a new connection occurs though